### PR TITLE
Finish Backend ꒰ᐢ. .ᐢ꒱₊˚⊹

### PR DESCRIPTION
**Pull request by chinosu**
**Data filtering/sanitising**
We go out of our way to remove any extra fields to prevent adding unexpected stuff to the database

**Data Validation**
Requests with invalid data (i.e., an empty password, a non-integer age) will now be rejected

**Hashed Passwords**
They're no longer in plain text on the database

**Refactoring**
We now have an abstract CustomError, a concrete BadRequestError and matching error handler middleware

### CODE REVIEW

File: backend/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+# Copy the custom prestart script
+COPY add-js-extension.sh .
+RUN chmod +x add-js-extension.sh
+
+# Build your TypeScript app
+RUN npm run build
+
+# Your app binds to port 3000 so you'll use the EXPOSE instruction to have it mapped by the docker daemon
+EXPOSE 8008
```
Comment (mark-trann) [inline backend/Dockerfile:24]: question: is this used to map the access of port 8008 from user to internal server's port 3000?

File: backend/package.json [hunk 2]
```diff
@@ -9,37 +9,34 @@
...
     "lint": "eslint .",
     "prettier": "prettier . --check",
-    "prettierfix": "prettier . --fix"
+    "prettierfix": "prettier . --write"
   },
   "author": "",
   "license": "ISC",
   "dependencies": {
     "bcrypt": "^5.1.1",
-    "cors": "^2.8.5",
```
Comment (mark-trann) [inline backend/package.json:21]: question: why'd we remove cors?

File: frontend/src/components/ProgressBar.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        className="
+          flex
+          items-center
+          justify-center
+          bg-indigo-600
+          rounded-full
+          transition-width
+          duration-300
+          ease-in-out"
+        style={{ width: `${progress}%`, height: '8px' }}
```
Comment (mark-trann) [inline frontend/src/components/ProgressBar.tsx]: can do this tailwind with `w-[${progress}%]` and `h-[8px]`

File: frontend/src/components/ProgressBar.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+          translate-x-1/2
+          transition-width
+          duration-300
+          ease-in-out"
+        style={{
+          width: '12px',
+          height: '12px',
+          top: '-2px',
+          left: `${Math.max(progress - 2, -1)}%`
+        }}
```
Comment (mark-trann) [inline frontend/src/components/ProgressBar.tsx]: similar to prev comment, can do this tailwind as well

File: frontend/src/components/ProgressBar.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
+// ProgressBar.js
+import { FC } from 'react';
+
+interface ProgressBarProps {
+    progress: number;
```
Comment (mark-trann) [inline frontend/src/components/ProgressBar.tsx]: indentation (run `npx prettier . -w`)

File: frontend/src/components/ProgressBar.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+          bg-indigo-600
+          rounded-full
+          transition-width
+          duration-300
+          ease-in-out"
+        style={{ width: `${progress}%`, height: '8px' }}
+      />
+      <div
+        className="
+          bg-indigo-600
```
Comment (mark-trann) [inline frontend/src/components/ProgressBar.tsx]: pls change color to fit the given figma (using defined color palette in our tailwind config)

File: frontend/src/components/ProgressBar.tsx [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+const ProgressBar: FC<ProgressBarProps> = ({ progress }) => {
+  return (
+    <div className="w-full bg-gray-300 rounded-full relative">
+      <div
+        className="
+          flex
+          items-center
+          justify-center
+          bg-indigo-600
```
Comment (mark-trann) [inline frontend/src/components/ProgressBar.tsx]: change color here as well pls

Comment (mark-trann) [reply frontend/src/components/ProgressBar.tsx]: hint: look at `CourseButton` component bg value

File: frontend/src/components/ProgressBar.tsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+          transform
+          translate-x-1/2
+          transition-width
+          duration-300
+          ease-in-out"
+        style={{
+          width: '12px',
+          height: '12px',
+          top: '-2px',
+          left: `${Math.max(progress - 2, -1)}%`
```
Comment (mark-trann) [inline frontend/src/components/ProgressBar.tsx]: the circle needs to be a bit more to the left to make it congruent with figma design
```suggestion
          left: `${Math.max(progress - 3, -1)}%`
```

**Review by mark-trann — COMMENTED** [review]
ty for the features and great work on the backend. pls review comments and lmk what u think!

File: frontend/src/components/ProgressBar.tsx (hunk 3 above)
Comment (chinosu) [reply frontend/src/components/ProgressBar.tsx]: but 
image
https://dev.to/alanscodelog/how-to-use-dynamic-variables-with-tailwind-4309

File: frontend/src/components/ProgressBar.tsx (hunk 8 above)
Comment (chinosu) [reply frontend/src/components/ProgressBar.tsx]: image

image

**Comment by chinosu** [comment]
@marktran2

**Review by mark-trann — APPROVED** [review]
ty for fixes - lgtm

File: frontend/src/components/ProgressBar.tsx (hunk 3 above)
Comment (mark-trann) [reply frontend/src/components/ProgressBar.tsx]: mmm kk interesting never realised this - no worries we can keep it like this 👍

**Pull request merged**